src/modbus.h

```diff
@@ -237,6 +246,11 @@ MODBUS_API int modbus_reply(modbus_t *ctx, const uint8_t *req,
                             int req_length, modbus_mapping_t *mb_mapping);
 MODBUS_API int modbus_reply_exception(modbus_t *ctx, const uint8_t *req,
                                       unsigned int exception_code);
+MODBUS_API int modbus_set_reply_callbacks(modbus_t *ctx,
+                                          const modbus_reply_callbacks_t *cb,
+                                          void *user_ctx);
+MODBUS_API int modbus_reply_callback(modbus_t *ctx, const uint8_t *req,
+                                     int req_length);
 
 /**
  * UTILS FUNCTIONS
```

Comment:
Some general notes:

1. this PR is very... big :smile: ... I'm not sure that I already understood all parts

1. When implementing a Modbus server, you usually want to do something with the data, i.e. the Modbus server is only one part of the whole application. The servers which I have written were multi-threaded, so you usually need some locking when accessing shared data. I think this can already prepared by adding lock/unlock callbacks.

1. I'm wondering whether extending the modbus context makes sense. As already stated in #407, I could image a second "layer" on top of the existing "ground or base layer" functions which cares about such higher-level functionality.

Reply:
1. Yes it's big, I think to make it easier to review I will split out the reply code in a separate file. I think it is not possible to further split the code in several commits. It's complex.

2. I agree with the fact that you need locking around the data, however I disagree that libmodbus has to take care of this within its API. If you need locking, just lock the `modbus_reply_callback()`-call in your code and access to the data is atomic.

3. Is the current reply-based-on-mapping-code used in anything else than examples and tests? When thinking about servers and slaves I was very quickly limited by the mapping-implementation, which made me develop this code. I my eyes, the reply-callback-code fits the library's style, no need to create extra layers.

I'm going to write/import examples in the test-dirs which will show how to write a multi-slave-server, that could help to see how to use it.

Thanks a lot for your review! :thumbsup:

Reply:
Can it be master slave opperate in modbus rtu in PIC18fXXX

Reply:
If libmodbus compiles and works for this platform, this callback-based server/slave should work as well.